### PR TITLE
Add/fix supplement_index_with_*

### PR DESCRIPTION
**Pull request by mcg1969**
This restores `add_unknown` functionality with a new function `supplement_index_with_cache`, and it simplifies `supplement_index_with_prefix` as well. It also moves `add_pip_dependency` into `get_index`.

### CODE REVIEW

File: conda/core/index.py [hunk 1]
```diff
@@ -47,39 +48,57 @@
...
+def supplement_index_with_prefix(index, prefix, channels):
+    # type: (Dict[Dist, IndexRecord], str, Set[canonical_channel]) -> None  # NOQA
     # supplement index with information from prefix/conda-meta
     assert prefix
-
-    priorities = {chnl: prrty for chnl, prrty in itervalues(channel_priority_map)}
-    maxp = max(itervalues(priorities)) + 1 if priorities else 1
+    maxp = len(channels) + 1
     for dist, info in iteritems(linked_data(prefix)):
-        fn = info['fn']
```
Comment (mcg1969) [inline conda/core/index.py]: This old dist string construction code should be removed, right? I mean, dist is available right there in `linked_data`.

**Comment by mcg1969** [comment]
@kalefranz the test fail was due to the fact that conda was adding cache packages to the index when in offline mode. I've modified the tests accordingly, testing both `unknown=True` and `unknown=False`. However, I've added behavior that changes the _default_ value of `unknown` from `False` to `True` when `context.offline` is `True`. After all, in offline mode, the only _installable_ packages are in the cache or local channels.

**Comment by mcg1969** [comment]
My point above is: you may not like that change, and if so, it's easy to back out, but the tests would have to be modified slightly again as well. If you _do_ like it, this will be ready to go as soon as the tests pass

**Comment by mcg1969** [comment]
OK, this is a significant simplification of explicit. In particular: _it never requires fetching `repodata.json` files_. Instead, each of the explicit URLs is put through the standard fetch process, and extracted. Once that is done, a "mini-index" consisting of their `index.json` files is constructed to facilitate the link and unlink process.

The one problem with `explicit` is that it doesn't use `display_actions` or provide for any sort of confirmation step.

**Comment by kalefranz** [comment]
This PR instead of https://github.com/conda/conda/pull/4108

**Pull request merged**